### ISSUE TITLE
Bulk inserts are slow (Slick 3 does individual inserts)

### ISSUE THREAD

**Issue by thesamet** [slick/slick#1272]
It turns out that when inserting using

```
table ++= Seq(v1,v2)
```

The SQL generated code (at least with the MySQL driver) generated individual INSERT statements, rather than a single INSERT that loads multiple values at once.

Demo project: https://github.com/thesamet/slick3-issue-1272

Output from MySQL query.log:

```
                 2675 Query     SET NAMES latin1
                 2675 Query     SET character_set_results = NULL
                 2675 Query     SET autocommit=1
                 2675 Query     SET sql_mode='STRICT_TRANS_TABLES'
                 2675 Query     create table `SUPPLIERS` (`SUP_ID` INTEGER NOT NULL PRIMARY KEY,`SUP_NAME` TEXT NOT NULL,`STREET` TEXT NOT NULL,`CITY
` TEXT NOT NULL,`STATE` TEXT NOT NULL,`ZIP` TEXT NOT NULL)
                 2675 Query     SET autocommit=0
                 2675 Query     insert into `SUPPLIERS` (`SUP_ID`,`SUP_NAME`,`STREET`,`CITY`,`STATE`,`ZIP`)  values (101,'Acme, Inc.','99 Market Stre
et','Groundsville','CA','95199')
                 2675 Query     insert into `SUPPLIERS` (`SUP_ID`,`SUP_NAME`,`STREET`,`CITY`,`STATE`,`ZIP`)  values (49,'Superior Coffee','1 Party Pl
ace','Mendocino','CA','95460')
                 2675 Query     insert into `SUPPLIERS` (`SUP_ID`,`SUP_NAME`,`STREET`,`CITY`,`STATE`,`ZIP`)  values (150,'The High Ground','100 Coffe
e Lane','Meadows','CA','93966')
                 2675 Query     commit
```

**Comment by szeiger** [comment]
Depending on the driver and other circumstances (like returning primary keys) Slick may fall back to individual inserts. In case of a simple insert on MySQL, it checks the JDBC Connection's DatabaseMetaData to determine if batch inserts are supported.

**Comment by brianwawok** [comment]
@thesamet i think you need to add `?rewriteBatchedStatements=true` to the DB url to make MySQL use batches....

**Comment by szeiger** [comment]
Marking as a documentation issue.

**Comment by stevorobs3** [comment]
I am having the same issue using the PostgresDriver. I use the ++= operator, expecting it to compile to a single insert but it does one insert per Values tuple.

Is there a solution for the PostgresDriver that you know about?

**Comment by brianwawok** [comment]
@stevorobs3  do you have a mini code example of what you are doing?

**Comment by stevorobs3** [comment]
hi @brianwawok yes, this is a rough snippet of what we are doing:

```scala
import scala.slick.driver.PostgresDriver.api._
val db: JDBCBackend.DatabaseDef = ???
...
case class SomeRow(rowNum: Int)
class SomeTable extends Table[SomeRow]("some_table") ...

db.run(TableQuery[SomeTable] ++= List.fill(100)(SomeRow(1)))
```

It seems that this generates 100 inserts of the form:
```sql
INSERT INTO some_table(row_num) values(1)
```

I read this [here](http://postgresql.nabble.com/Resurrected-thread-Speed-improvement-Group-batch-Insert-Rewrite-the-INSERT-at-the-driver-level-using-td5843472.html), but I'm not 100% sure whether that means its not supported or if there is a different work around.

**Comment by brianwawok** [comment]
@stevorobs3 and how do you know these 100 inserts are going as 100 trips to the database, instead of 1 big batch of 100 inserts?  In the MYSQL above it was super obvious.. like you could only insert 1k records per second (database latency was about 1 MS). After the driver param switch, you could magically insert 40k records per second.

**Comment by stevorobs3** [comment]
@brianwawok I looked at the postgres logs and saw the queries being logged out. Yes, they may well have been sent in a single batch but in was 100 individual inserts rather than one of the form:
```sql
INSERT INTO some_table(row_num) values (1),(1),(1),(1),(1),(1),(1),(1),(1) ...etc
```

We converted this to using the sqlu interpolator to the above code and it ran ~20 times faster even on inserting ~400 data items at a time.

**Comment by adevore3** [comment]
@brianwawok We're seeing the same thing with postgres when trying to insert a batch of 1000. We couldn't figure out how to get slick to log the insert statement(s) it was making but we noticed the database logged individual inserts.

**Comment by brianwawok** [comment]
@adevore3 could be a bug, I don't have a working copy of slick nor postgres right now ;(  May be worth making a new issue for with the info you have?

**Comment by adevore3** [comment]
@brianwawok 
-      val insertQuery = alertsSingleton returning alertsSingleton.map(_.id) into ((alert, id) => alert.copy(id = id))
-      recordFutureFailures(db.run { insertQuery ++= alerts })
+      recordFutureFailures(db.run { alertsSingleton ++= alerts} .map( a => Unit ))

It turns out that the operator does work if you do it on the table object. Our problem was that we had a query that tried to return the id into the object, we then tried to use that query with the batch operator which resulted in N insert statements instead of 1 insert statement.

To verify we tried inserting a batch of 100,000. The old way timed out after 10 minutes and only inserted about 40,000. The new way inserted all 100,000 in less than 13 seconds. 

tl;dr The batch insert operator works but beware of the query you run it against.

Hope this helps anyone in the future.

**Comment by d6y** [comment]
FYI: regarding Postgres, there's a PR open in the Postgres JDBC driver to detect batch inserts and re-write them: https://github.com/pgjdbc/pgjdbc/pull/491

**Comment by davecramer** [comment]
@d6y can you test this the pgjdbc PR with slick ?

**Comment by d6y** [comment]
@davecramer I've done some quick testing with a local `mvn package -DskipTests` of [the batch-rewite-03 branch](https://github.com/whitingjr/pgjdbc/commits/batch-rewrite-03). Providing I modify my JDBC connection to use `?reWriteBatchedInserts=true` and run my insert in a transaction I do see a re-write in the _postgres-server.log_ file:

```
LOG:  execute <unnamed>: BEGIN
LOG:  execute <unnamed>: insert into "messagenopk" ("sender","content","ts")  values ($1,$2,$3),($4,$5,$6),($7,$8,$9),($10,$11,$12),($13,$14,$15),($16,$17,$18)
...
```

**Comment by davecramer** [comment]
@d6y Thanks!

**Comment by jeantil** [comment]
On mysql at least there seems to be one side effect to using rewriteBatchedStatements : the number of inserted elements is no longer returned. 

```
package eu.byjean

import scala.concurrent.{ Await, Future }
import scala.concurrent.duration._
import scala.concurrent.ExecutionContext.Implicits._

import com.typesafe.config.{ Config, ConfigFactory }
import slick.driver.MySQLDriver
import slick.driver.MySQLDriver.api._
import slick.profile.SqlAction

object BatchInsert extends App {
  private val settings =
    """
      |db {
      |  driver = com.mysql.jdbc.Driver
      |  url = "jdbc:mysql://localhost:3306/rzv2?&useSSL=false&rewriteBatchedStatements=false"
      |  keepAliveConnection = true
      |  connectionPool = disabled
      |  databaseName = "jean"
      |  user = "jean"
      |  password = "jean"
      |}
    """.stripMargin
  private val config: Config = ConfigFactory.parseString(settings)
  private val db: MySQLDriver.backend.Database = MySQLDriver.backend.Database.forConfig("db", config)

  class Coffees(tag: Tag) extends Table[(String, Int)](tag, "COFFEES") {
    def name = column[String]("COF_NAME", O.PrimaryKey)
    def supID = column[Int]("SUP_ID")
    def * = (name, supID)
  }
  private val coffees = TableQuery[Coffees]

  private def bulkInsert = {
    val newCoffees = (1 to 100).map(i => ("name-" + i, i))
    val insert = coffees ++= newCoffees
    db.run(insert)
  }

  val x = for {
    create <- db.run(coffees.schema.create)
    insert <- bulkInsert
    drop <- db.run(coffees.schema.drop)
    _ = println(insert)
  } yield insert
  Await.result(x, 30.seconds)

}
```

Prints :
* `None`  when `rewriteBatchedStatements=true`
* `Some(100)`  when `rewriteBatchedStatements=false`

**Comment by Proximator** [comment]
I am trying to use slick 3 bulk inserts using sql server and performance are very bad, about 4 times slower than JDBC. 
Please refer to : stackoverflow.com/questions/41262179/slick-3-0-bulk-inserts-are-slow-with-sql-server

**Comment by nikkatalnikov** [comment]
Confirm @Proximator 
NO bulk inserts executed with com.microsoft.sqlserver.jdbc.SQLServerDriver

**Comment by code33** [comment]
in my dao :
```

  def f[T](v: T) = v match {
    case _: Int    => "Int"
    case _: String => "String"
    case _         => "Unknown"
  }

  def insert2Vals(rows:Seq[(Any,Any)]):String={
      s"""
        ${rows.map(ite => s"(" +
         s"${if(f(ite._1).equals("String"))  s"'${ite._1}'" else ite._1  }," +
         s"${if(f(ite._2).equals("String"))  s"'${ite._2}'" else ite._2  }"  +
         s")").mkString(",")}

      """
  }

     val beInsert = rows.map(r=>{
         (r.uSn,r.miSn)
      })

      val sql = sql"""
          insert into relation_msg_user_t (u_sn,mi_sn) values #${insert2Vals(beInsert)} returning rmu_id
      """.as[Int]

      db.run(sql)


```

不用谢

**Comment by code33** [comment]
> in my dao :
> 
> ```
> 
>   def f[T](v: T) = v match {
>     case _: Int    => "Int"
>     case _: String => "String"
>     case _         => "Unknown"
>   }
> 
>   def insert2Vals(rows:Seq[(Any,Any)]):String={
>       s"""
>         ${rows.map(ite => s"(" +
>          s"${if(f(ite._1).equals("String"))  s"'${ite._1}'" else ite._1  }," +
>          s"${if(f(ite._2).equals("String"))  s"'${ite._2}'" else ite._2  }"  +
>          s")").mkString(",")}
> 
>       """
>   }
> 
>      val beInsert = rows.map(r=>{
>          (r.uSn,r.miSn)
>       })
> 
>       val sql = sql"""
>           insert into relation_msg_user_t (u_sn,mi_sn) values #${insert2Vals(beInsert)} returning rmu_id
>       """.as[Int]
> 
>       db.run(sql)
> ```
> 
> 不用谢

and u could be custom insert3Vals insert4Vals and more vals sql contact function

**Comment by Bugsource** [comment]
> @brianwawok I looked at the postgres logs and saw the queries being logged out. Yes, they may well have been sent in a single batch but in was 100 individual inserts rather than one of the form:
> 
> ```sql
> INSERT INTO some_table(row_num) values (1),(1),(1),(1),(1),(1),(1),(1),(1) ...etc
> ```
> 
> We converted this to using the sqlu interpolator to the above code and it ran ~20 times faster even on inserting ~400 data items at a time. I just expected that there would be a way for the `++=` operator to directly produce this SQL code.

Make sense！So the slick have't solve this after all these years. Disappointed.

**Comment by nafg** [comment]
A pull request would be very welcome

**Comment by slovvik** [comment]
Today I had the same problem. It seems that you can't map the result of the batch insert. I have tested it using slick SQL query logger `<logger name="slick.jdbc.JdbcBackend.statement" level="DEBUG"/>`.
Here are examples:
```scala
//Working batch insert
coffees ++= coffeesList

//Not working batch insert
(coffees returning coffees.map(_.id)) ++= coffeesList
```

**Comment by nafg** [comment]
Would you like to contribute an implementation?

**Comment by slovvik** [comment]
I would like to, but I think it is too hard for me.

**Comment by jkugiya** [comment]
I feel like I could implement this, but on the other hand, I think we should separate the APIs because sometimes the existing implementation is more reasonable for the application. There can be systems where you have to choose between either method depending on the situation.
For example, there are cases where adding a very large number of rows would exceed the buffer size of the connection, and it would be desirable to issue SQL in a split manner.

For example, I'm thinking of an API like the following

1. 
````scala
// batch insert(multiple single insert)
coffees ++= cofeesList
// bulk insert
coffees +++= cofeesList
````

2. 
```scala
// batch insert(multiple single insert)
coffees ++= cofeesList
// bulk insert
coffees.insertBulk(cofeesList)
```

**Comment by nafg** [comment]
I would not introduce new operators without a very strong need

**Comment by jkugiya** [comment]
https://github.com/jkugiya/slick/pull/1/files

This seems to be working, but I'm changing an API that I don't really understand(`ResultConverter`) , so I'm trying to figure out if this is the best way to go.

**Comment by jkugiya** [comment]
#2398 
I have implemented this feature. This implementation can probably be reused for `insertOrUpdate`, but we need to come up with a good naming convention to make it coexist with the existing(batch insert) implementation.
